README
INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install
   or
   perl Build.pl --cgi=/srv/www/cgi-bin --htd=/srv/www/htdocs --skip
   ./Build && ./Build test && ./Build install

This module requires these other modules and libraries:

  CGI::LZE                     >= 0.25,
  HTML::Template::LZE          >= 0.25,
  HTML::Template::LZE::Window  >= 0.25,
  HTML::Entities               >= 1.35,
  Syntax::Highlight::Perl      >= 1.00,

COPYRIGHT AND LICENCE

Copyright (C) 2006-2007 by Hr. Dirk Lindner 

This program is free software; you can redistribute it and/or modify it under the terms of the GNU Lesser General Public License as published by the Free Software Foundation; This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more details. 
 


Changes
Revision history for Perl extension HTML::Template::LZE::Editor 0.26

0.26
        Overwrought Docs and clean up code, fix build System

0.25
        Clean Build system
0.24
        workaround
0.23
        use Module Build, new namespace  HTML::Template::LZE::Editor


examples/editor.pl
#!/usr/bin/perl -w
use lib("lib");
use HTML::Template::LZE::Editor;
use HTML::LZE::BBCODE;
use CGI::LZE qw(:all);
init("/home/groups/l/li/lindnerei/cgi-bin/config/settings.pl");
print header;
print start_html(-title => 'HTML::Template::LZE::Editor', -script => [{-type => 'text/javascript', -src => '/javascript/editor.js'}], -style => '/style/Crystal/editor.css',);
if(param('action') && param('action') eq 'add') {
        my $txt = param('message');
        if(param('submit') eq translate('preview')) {
                print a({href => "$ENV{SCRIPT_NAME}?txt=$txt"}, 'Edit it Again');
                print h2(param('headline'));
                BBCODE(\$txt);
                print br(), $txt;
        } else {
                print "You should save it ", br();
                print $txt;
                print br(), a({href => "$ENV{SCRIPT_NAME}?txt=$txt"}, 'Edit it Again');
        }
} else {
        my %parameter = (

                action => 'add',

                body => param('txt') ? param('txt') : 'body of the message',

                class => "min",

                attach => 0,

                maxlength => '100',

                path => "/home/groups/l/li/lindnerei/cgi-bin/templates/",

                reply => '',

                server => "http://localhost",

                style => 'Crystal',

                thread => 'news',

                headline => "New Message",
                title    => "title",
                catlist  => ' ',
                html     => 0,               # html enabled ? 0 for bbcode
        );
        my $editor = new HTML::Template::LZE::Editor(\%parameter);

        print $editor->show();
}
use showsource;
&showSource("./editor.pl");


lib/HTML/LZE/BBCODE.pm
package HTML::LZE::BBCODE;
use strict;
use warnings;
use vars qw(@EXPORT @ISA );
require Exporter;
@HTML::LZE::BBCODE::EXPORT  = qw(BBCODE);
@ISA                        = qw(Exporter);
$HTML::LZE::BBCODE::VERSION = '0.26';
use HTML::Entities;

=head1 NAME

HTML::LZE::BBCODE

=head1 required Modules

HTML::Entities,

Syntax::Highlight::Perl

=head1 SYNOPSIS

use HTML::LZE::BBCODE;

my $test = '

[code]

print "testIt";

[/code]

';

BBCODE(\$test);

print $test;

=head1 DESCRIPTION

bbcode to html

=head2 EXPORT

BBCODE()

=head2 BBCODE

=cut

=head1 Public

=head2 BBCODE

=cut

sub BBCODE {
        my $string = shift;
        my $eval   = shift;
        $eval = defined $eval ? $eval : 0;
        my $rplc;
        if($$string =~ /\[code\](.*?)\[\/code\]/gs) {
                use Syntax::Highlight::Perl ':FULL';
                my $color_Keys = {
                                  'Variable_Scalar'   => 'Variable_ # or Scalar',
                                  'Variable_Array'    => 'Variable_Array',
                                  'Variable_Hash'     => 'Variable_Hash',
                                  'Variable_Typeglob' => 'Variable_Typeglob',
                                  'Subroutine'        => 'Subroutine',
                                  'Quote'             => 'Quote',
                                  'String'            => 'String',
                                  'Comment_Normal'    => 'Comment_Normal',
                                  'Comment_POD'       => 'Comment_POD',
                                  'Bareword'          => 'Bareword',
                                  'Package'           => 'Package',
                                  'Number'            => 'Number',
                                  'Operator'          => 'Operator',
                                  'Symbol'            => 'Symbol',
                                  'Character'         => 'Character',
                                  'Directive'         => 'Directive',
                                  'Label'             => 'Label',
                                  'Line'              => 'Line',
                };
                my $formatter = new Syntax::Highlight::Perl;
                $formatter->define_substitution('<' => '&lt;', '>' => '&gt;', '&' => '&amp;',);    # HTML escapes.
                while(my ($type, $style) = each %{$color_Keys}) {
                        $formatter->set_format($type, [qq|<span class="$style">|, '</span>']);
                }
                my $perldoc_Keys = {'Builtin_Operator' => 'Builtin_Operator', 'Builtin_Function' => 'Builtin_Function', 'Keyword' => 'Keyword',};
                while(my ($type, $style) = each %{$perldoc_Keys}) {
                        $formatter->set_format($type, [qq|<a onclick="window.open('http://perldoc.perl.org/search.html?q='+this.innerHTML)" class="$style">|, "</a>"]);
                }
                $rplc = $formatter->format_string($1);
                $rplc    =~ s?\n?<br/>?g;
                $$string =~ s/\[code\](.*?)\[\/code\]/\[Formatstring\/\]/gs;
        }
        if($eval > 4) {
                my $return;
                if($$string =~ /\[eval\](.*?)\[\/eval\]/) {
                        $return = eval $1;
                }
                $$string =~ s/\[eval\](.*?)\[\/eval\]/$return/gs;
        }
        $$string = encode_entities($$string);
        $$string =~ s:\[(u)\](.*?)\[/\1\]:<$1>$2</$1>:gs;
        $$string =~ s:\[(b)\](.*?)\[/\1\]:<$1>$2</$1>:gs;
        $$string =~ s/\[(b)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(i)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(ol)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(ul)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(li)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(h1)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(h2)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(h3)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(h4)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(h5)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(s)\]([^\[\/\1\]]*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(sub)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[(sup)\](.*?)\[\/\1\]/<$1>$2<\/$1>/gs;
        $$string =~ s/\[hr\]/<hr\/>/gs;
        $$string =~ s/\[color=(.*?)\](.*?)\[\/color\]/<span style="color:$1;background-color:#E6DADE;">$2<\/span>/gs;
        $$string =~ s/\[right\](.*?)\[\/right\]/<div align="right">$1<\/div>/gs;
        $$string =~ s/\[center\](.*?)\[\/center\]/<div align="center">$1<\/div>/gs;
        $$string =~ s/\[left\](.*?)\[\/left\]/<div align="left">$1<\/div>/gs;
        $$string =~ s/\[url=(.*?)\](.*?)\[\/url\]/<a style="color:#000000;background-color:#E6DADE;" href="$1">$2<\/a>/gs;
        $$string =~ s/\[email=(.*?)\](.*?)\[\/email\]/<a style="color:#000000;" target="_blank" href="mailto:$1">$2<\/a>/gs;
        $$string =~ s/\[(img)\](.*?)\[\/\1\]/<img border="0" src="$2" alt=""\/>/g;
        $$string =~ s/\[(google)\](.*?)\[\/\1\]/<a style="color:#000000;" target="_blank" href="http:\/\/www.google.de\/search?q=$2">Google:$2<\/a>/gs;
        $$string =~
          s/\[blog=(.*?)\](.*?)\[\/blog\]/<table  cellpadding="0" cellspacing="0" border="0"><tr><td><table  cellpadding="5" cellspacing="0"  border="0" class="blog"><tr><td>$2<\/td><\/tr><\/table><\/td><\/tr><tr><td><b><a href="$1" class="link">Quelle<\/a><\/b><\/td><\/tr><\/table>/gs;
        $$string =~ s/\n/\n<br\/>/ig;
        $$string =~ s/:\)/<img src="\/images\/smiley.gif" alt=":)" border="0"\/>/g;
        $$string =~ s/;D/<img src="\/images\/grin.gif" alt=";D" border="0"\/>/g;
        $$string =~ s/8\)/<img src="\/images\/cool.gif" alt="8)" border="0"\/>/g;
        $$string =~ s/:-\*/<img src="\/images\/kiss.gif" alt=":-*" border="0"\/>/g;
        $$string =~ s/:\(/<img src="\/images\/angry.gif" alt=":(" border="0"\/>/g;
        $$string =~ s/:\(/<img src="\/images\/angry.gif" alt=":(" border="0"\/>/g;
        $$string =~ s/\[Formatstring\/\]/$rplc/gs;
}

=head1 AUTHOR

Dirk Lindner <lindnerei@o2online.de>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Hr. Dirk Lindner

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public License
as published by the Free Software Foundation; 
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

=cut

1;



lib/HTML/Template/LZE/Editor.pm
package HTML::Template::LZE::Editor;
use HTML::Template::LZE::Window;
use CGI::LZE qw(translate);
use strict;
use warnings;
use vars qw(
  $catlist
  $class
  $DefaultClass @EXPORT  @ISA
  $path
  $right
  $server
  $style
  $title
  $body
  $maxlength
  $action
  $reply
  $thread
  $headline
  $html
);
require Exporter;
use HTML::Template::LZE;
@HTML::Template::LZE::Editor::ISA         = qw( Exporter HTML::Template::LZE);
@HTML::Template::LZE::Editor::EXPORT_OK   = qw(initEditor show );
%HTML::Template::LZE::Editor::EXPORT_TAGS = ('all' => [qw(initEditor show )]);

$HTML::Template::LZE::Editor::VERSION = '0.26';

$DefaultClass = 'HTML::Template::LZE::Editor' unless defined $HTML::Template::LZE::Editor::DefaultClass;

=head2 new()

=cut

sub new {
        my ($class, @initializer) = @_;
        my $self = {};
        bless $self, ref $class || $class || $DefaultClass;
        $self->initEditor(@initializer) if(@initializer);
        return $self;
}

=head2 initEditor()

       my %parameter =(

       action   = > 'action',

       body     => 'body of the message',

       class    => "min",

       attach   => '1',

       maxlength => '100',

       path   => "/srv/www/cgi-bin/templates",#default : '/srv/www/cgi-bin/templates'

       reply    =>  '', #default : ''

       server   => "http://localhost", #default : 'http://localhost'

       style    =>  $style, #default : 'Crystal'

       thread   =>  'news',#default : ''

       headline    => "&New Message", #default : 'headline'

       html     => 1 , # html enabled ? 0 for bbcode default : 0

       text     => 'the body', #default : 'headline'

       );

       my $editor = new HTML::Template::LZE::Editor(\%parameter);

       print $editor->show();

=cut

sub initEditor {
        my ($self, @p) = getSelf(@_);
        my $hash = $p[0];
        $server    = defined $hash->{server}    ? $hash->{server}    : 'http://localhost';
        $style     = defined $hash->{style}     ? $hash->{style}     : 'Crystal';
        $title     = defined $hash->{title}     ? $hash->{title}     : 'Editor';
        $path      = defined $hash->{path}      ? $hash->{path}      : '/srv/www/cgi-bin/templates';
        $body      = defined $hash->{body}      ? $hash->{body}      : 'Text';
        $maxlength = defined $hash->{maxlength} ? $hash->{maxlength} : '300';
        $action    = defined $hash->{action}    ? $hash->{action}    : 'addMessage';
        $reply     = defined $hash->{reply}     ? $hash->{reply}     : '';
        $thread    = defined $hash->{thread}    ? $hash->{thread}    : 'news';
        $headline  = defined $hash->{headline}  ? $hash->{headline}  : 'headline';
        $catlist   = defined $hash->{catlist}   ? $hash->{catlist}   : '';
        $right     = defined $hash->{right}     ? $hash->{right}     : 0;
        $html      = $hash->{html}              ? $hash->{html}      : 0;
        $class = 'min' unless (defined $class);
        my %template = (path => $hash->{path}, style => $style, template => 'editor.html',);
        $self->SUPER::initTemplate(\%template);
}

=head2 show()

=cut

sub show {
        my ($self, @p) = getSelf(@_);
        $self->initEditor(@p) if(@p);
        my %parameter = (path => $path, style => $style, title => $title, server => $server, id => 'winedit', class => $class,);
        my $output    = '<br/>';
        my $window2   = new HTML::Template::LZE::Window(\%parameter);
        $output .= $window2->windowHeader();
        my $att = ($right >= 2) ? '<tr><td align="center">' . translate('chooseFile') . ':<input name="file" type="file" accept="text/*" maxlength="2097152" size ="30" title="Datei zum Hochladen ausw&#228;hlen"/></td></tr>' : "";
        my %editor = (
                      name      => 'editor',
                      server    => $server,
                      style     => $style,
                      title     => $title,
                      body      => $body,
                      maxlength => $maxlength,
                      action    => $action,
                      reply     => $reply,
                      thread    => $thread,
                      headline  => $headline,
                      catlist   => $catlist,
                      attach    => $att,
                      html      => $html
        );
        $output .= $self->SUPER::appendHash(\%editor);
        $output .= $window2->windowFooter();
        return $output;

}

=head2 getSelf()

=cut

sub getSelf {
        return @_ if defined($_[0]) && (!ref($_[0])) && ($_[0] eq 'HTML::Template::LZE::Editor');
        return (defined($_[0]) && (ref($_[0]) eq 'HTML::Template::LZE::Editor' || UNIVERSAL::isa($_[0], 'HTML::Template::LZE::Editor'))) ? @_ : ($HTML::Template::LZE::Editor::DefaultClass->new, @_);
}

=head1 AUTHOR

Dirk Lindner <lindnerei@o2online.de>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 - 2008 by Hr. Dirk Lindner

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public License
as published by the Free Software Foundation; 
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

=cut

1;


Build.PL
use strict;
use Module::Build;
use vars qw( $settings $skip);
my $class = Module::Build->subclass(class => "Module::Build::Custom", code => <<'SUBCLASS' );
my $release = 'blib/rhtml/';
my $dir     = 'htdocs/';
my $path = '.';
sub ACTION_build {
        my $self = shift;
        &change($dir);
        $release = 'blib/perl/';
        $dir     = 'cgi-bin/';
        &change($dir);
}
sub change {
        my $d = shift;
        chomp($d);
        opendir(IN, $d) or warn "cant open $d $!:$/";
        my @files = readdir(IN);
        closedir(IN);
        for(my $i = 0 ; $i <= $#files ; $i++) {
                unless ($files[$i] =~ /^\./) {
                        my $c = $d . $files[$i];
                        my $e = $c;
                        $e =~ s/^$dir(.*)/$1/;
                        unless (-d $d . $files[$i]) {
                                system(" cp " . $c . " $release/$e") unless (-e $release . "/" . $e && $files[$i] =~ /\~$/);
                        } else {
                                system("mkdir -p $release/$e") unless (-e $release . "/" . $e);
                                &change($d . $files[$i] . "/");
                        }
                }
        }
}
SUBCLASS
my $inctance = $class->new(
        get_options => {skip => {store => \$skip},},
        module_name => 'HTML::Template::LZE::Editor',
        requires => {
            'Test::More' => ">=0.7",
            'Module::Build' => ">=0.2808",
            'Test::Pod' =>  ">=1.26",
            'Test::Pod::Coverage' =>  ">=0.55",
            'CGI::LZE' => '>=0.25',
            'HTML::Template::LZE'  => '>=0.25',
            'HTML::Template::LZE::Window'  => '>=0.25',
            'HTML::Entities' => ">=1.35",
            'Syntax::Highlight::Perl' => ">=1.00",
},
($] >= 5.005 ? (AUTHOR => 'Dirk Lindner <lze@cpan.org>') : ()),
dist_abstract => 'BBCode & html editor',
license => 'lgpl',
create_makefile_pl => 'passthrough',
);
$inctance->{properties}{htd} = "/srv/www/htdocs";
$inctance->{properties}{cgi}  = "/srv/www/cgi-bin";

eval 'use CGI::LZE::Settings qw($settings loadSettings saveSettings)';
unless ($@) {
        loadSettings();
        *settings    = \$CGI::LZE::Settings::settings;
        $inctance->{properties}{htd}          = $settings->{cgi}{DocumentRoot} if -e $settings->{cgi}{DocumentRoot};
        $inctance->{properties}{cgi}         = $settings->{cgi}{bin} if -e $settings->{cgi}{bin};
}
$inctance->{properties}{cgi}         = $inctance->args('cgi')         if $inctance->args('cgi');
$inctance->{properties}{htd}         = $inctance->args('htd')         if $inctance->args('htd');
unless ($skip){
        $inctance->{properties}{htd}  = Module::Build->prompt("Where is your Document root ?",     $inctance->{properties}{htd} );
         $inctance->{properties}{cgi}  = Module::Build->prompt("Where is your cgi-bin Directory ?",$inctance->{properties}{cgi});
}else{
        print "Document root: $inctance->{properties}{htd}$/cgi-bin Directory:$inctance->{properties}{cgi}$/";
}
$inctance->{properties}{install_path}{perl} = $inctance->{properties}{cgi};
$inctance->{properties}{install_path}{rhtml} = $inctance->{properties}{htd};
$inctance->create_build_script;
